ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp

```diff
@@ -0,0 +1,164 @@
+//*****************************************************************************
+// Copyright 2017-2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <map>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ONNX_NAMESPACE
+{
+    class GraphProto;
+    class NodeProto;
+    class ValueInfoProto;
+} // namespace ONNX_NAMESPACE
+
+namespace ngraph
+{
+    enum class EdgeType
+    {
+        INPUT,
+        OUTPUT
+    };
+
+    template <EdgeType>
+    struct Edge
+    {
+        Edge() = delete;
+        Edge(const int node_idx, std::string tensor_name)
+            : m_node_idx{node_idx}
+            , m_tensor_name{std::move(tensor_name)}
+        {
+        }
+
+        const int m_node_idx;
+        const std::string m_tensor_name;
+    };
+    namespace onnx_import
+    {
+        /// \brief Defines an edge connected to an input of any node in the graph.
+        ///        It consists of a node index in the processed ONNX model and the input name.
+        ///        The index should point to a node in the topological sort of the underlying graph
+        ///        which means it has to be in range:  0 <= node_idx < graph.node_size()
+        ///
+        ///        For a node number 5, with 3 inputs:
+        ///
+        ///            ----(in_A)---->  +--------+
+        ///            ----(in_B)---->  | node 5 |  ----(out)---->
+        ///            ----(in_C)---->  +--------+
+        ///
+        ///        there are 3 possible valid instances of this struct:
+        ///            InputEdge(5, "in_A")
+        ///            InputEdge(5, "in_B")
+        ///            InputEdge(5, "in_C")
```

Comment:
Assuming that `node 5` is an operation and `(in_A)` etc. are tensors, this documentation doesn't provide rationale why we are specifying node index, because there is only one consumer for each of the tensors on the schematics. In this case the tensor name should be enough to specify the cutting point.

Reply:
As discussed offline, this will be addressed in a separate ticket: 47578